File: scala/core/src/main/scala/blockchain/ledger/ValueCalculator.scala
package blockchain.ledger

import blockchain.models.{GraphEntry, TransactionOutput}
import cats.effect.{Resource, Sync}
import cats.implicits.*
import com.google.protobuf.struct

trait ValueCalculator[F[_]] {
  def requiredMinimumQuantity(output: TransactionOutput): F[Long]
}

class ValueCalculatorImpl[F[_]: Sync] extends ValueCalculator[F] {
  // TODO: Add these numbers to ProtocolSettings
  def requiredMinimumQuantity(output: TransactionOutput): F[Long] =
    List(
      Sync[F].delay(100L), // dust prevention
      Sync[F].delay(output.account.as(50L).getOrElse(0L)), // association with an existing staking account
      Sync[F].delay(output.value.accountRegistration.as(1000L).getOrElse(0L)), // new staker registration
      Sync[F].delay(output.value.graphEntry.fold(0L)(graphEntryMinimumQuantity)) // graph data
    ).sequence.map(_.sum)

  def graphEntryMinimumQuantity(graphEntry: GraphEntry): Long =
    graphEntry.entry match {
      case GraphEntry.Entry.Vertex(v) => v.label.length.toLong * 10 + v.data.fold(0L)(protoStructMinimumQuantity)
      case GraphEntry.Entry.Edge(e)   => e.label.length.toLong * 10 + 100L + e.data.fold(0L)(protoStructMinimumQuantity)
      case _                          => 0L
    }

  def protoValueMinimumQuantity(value: struct.Value): Long =
    value.kind match {
      case struct.Value.Kind.StringValue(v) => v.length.toLong * 10
      case struct.Value.Kind.NumberValue(v) => v.toString.length.toLong * 10
      case struct.Value.Kind.BoolValue(_)   => 10L
      case struct.Value.Kind.ListValue(v)   => v.values.map(protoValueMinimumQuantity).sum
      case struct.Value.Kind.StructValue(v) => protoStructMinimumQuantity(v)
      case _                                => 10L
    }

  def protoStructMinimumQuantity(value: struct.Struct): Long =
    value.fields.toList.map(f => f._1.length + protoValueMinimumQuantity(f._2)).sum
}

object ValueCalculatorImpl:
  def make[F[_]: Sync]: Resource[F, ValueCalculator[F]] =
    Resource.pure(new ValueCalculatorImpl[F])

File: rust/giraffe_sdk/src/ledger/value_calculator.rs
use prost_types::Struct;

use crate::models::TransactionOutput;

pub fn required_minimum_quantity(output: &TransactionOutput) -> u64 {
    let mut res: u64 = 100;
    if output.asset.is_some() {
        res += 100;
    }
    if let Some(label) = &output.label {
        res += label.len() as u64 * 10;
    }
    if let Some(data) = &output.data {
        res += data_minimum_quantity(&data);
    }
    if output.edge.is_some() {
        res += 100;
    }
    if output.staking_registration.is_some() {
        res += 1000;
    }
    res
}

fn data_minimum_quantity(value: &Struct) -> u64 {
    let mut res: u64 = 0;
    for (key, value) in value.fields.iter() {
        res += key.len() as u64 + value_minimum_quantity(value);
    }
    res
}

fn value_minimum_quantity(value: &prost_types::Value) -> u64 {
    match value.kind.as_ref().unwrap() {
        prost_types::value::Kind::NumberValue(n) => n.to_string().len() as u64 * 10,
        prost_types::value::Kind::StringValue(s) => s.len() as u64 * 10,
        prost_types::value::Kind::StructValue(s) => data_minimum_quantity(&s),
        prost_types::value::Kind::ListValue(l) => {
            l.values.iter().map(|v| value_minimum_quantity(v)).sum()
        }
        _ => 10,
    }
}
